### PR TITLE
UI: line chart

### PR DESCRIPTION
**Pull request by DingoEatingFuzz**
_This is parter of a larger project to improve the cpu and memory usage charts in the UI_

Just a couple line chart components.

**Features**

1. Chart will always assume the dimensions of the container element, making this really flexible
2. The y-axis ticks count is dynamic based on the height of the chart, but will always be an odd number to ensure a line at the midpoint.
3. Charts resize automatically when the browser resizes.
4. A tooltip in the style of the distribution bar tooltip
5. Auto updates when data is added to the bound data property
6. The tooltip also auto updates as data comes in and shifts the datum that's underneath the mouse
7. A use-case specific stats chart that guarantees at least a five minute window instead of stretching 10 seconds of data across the entire width of the chart
8. A hip gradient fill that has a "global" intensity instead of a "local" intensity (explained below).

image

_Tooltip in action_

image

_Some smaller charts, with less y-axis ticks_

image

_A demonstration of the "global intensity" of the gradient. Notice that the chart on the right doesn't have as deep as blues near the line. This is because the intensity is mapped from the top to the bottom of the canvas instead of from the top to the bottom of the area under the line for a chart._

### CODE REVIEW

File: ui/app/templates/components/line-chart.hbs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    </linearGradient>
+    <clipPath id="{{maskId}}">
+      <path class="fill" d="{{area}}" />
+    </clipPath>
+  </defs>
+  <g class="y-gridlines gridlines" transform="translate({{yAxisOffset}}, 0)"></g>
+  <g class="canvas {{chartClass}}">
+    <path class="line" d="{{line}}" />
+    <rect class="area" x="0" y="0" width="{{yAxisOffset}}" height="{{xAxisOffset}}" fill="url(#{{fillId}})" clip-path="url(#{{maskId}})" />
+    <rect class="hover-target" x="0" y="0" width="{{yAxisOffset}}" height="{{xAxisOffset}}" onhover={{action updateActiveDatum}} />
```
Comment (meirish) [inline ui/app/templates/components/line-chart.hbs]: Why both the onhover here and the mouseenter/move in the d3 binding in the JS?

File: ui/tests/unit/components/stats-time-series-test.js [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+  });
+});
+
+test('x scale domain is at least five minutes', function(assert) {
+  const chart = this.subject();
+
+  chart.set('data', narrowData);
+
+  assert.equal(
+    +chart.get('xScale').domain()[0],
```
Comment (meirish) [inline ui/tests/unit/components/stats-time-series-test.js]: Is this just coercing to a Number?

File: ui/app/templates/components/line-chart.hbs (hunk 1 above)
Comment (DingoEatingFuzz) [reply ui/app/templates/components/line-chart.hbs]: Oh, oops. This onhover in the template should be deleted. Glimmer doesn't support binding events on svg elements; at least not in 2.18.

File: ui/app/templates/components/line-chart.hbs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  </g>
+  <g class="x-axis axis" transform="translate(0, {{xAxisOffset}})"></g>
+  <g class="y-axis axis" transform="translate({{yAxisOffset}}, 0)"></g>
```
[the remark this answers is not in the document]
Comment (DingoEatingFuzz) [reply ui/app/templates/components/line-chart.hbs]: Oh neat! Yeah, I'll look into this. The stat line charts, once they are in place, will also have related graphics and stats around them for the current value. I think those places will be the most useful for accessibility, but that doesn't mean svgs shouldn't have descriptions at least.

File: ui/tests/unit/components/stats-time-series-test.js (hunk 2 above)
Comment (DingoEatingFuzz) [reply ui/tests/unit/components/stats-time-series-test.js]: Yep. Comparing dates doesn't work due to them being passed by reference.

File: ui/app/templates/components/line-chart.hbs (hunk 3 above)
[the remark this answers is not in the document]
Comment (DingoEatingFuzz) [reply ui/app/templates/components/line-chart.hbs]: After playing with this for an hour or so I think the amount of work justifies its own PR. I'd still open it against the `f-ui-improved-stats-chart` branch, so it'd be done before the line chart hits `master`.

[the remark this answers is not in the document]
Comment (meirish) [reply ui/app/templates/components/line-chart.hbs]: Ah yeah that makes sense! Thanks for digging into it ❤️

File: ui/app/components/line-chart.js [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+    return d3Axis
+      .axisRight()
+      .scale(this.get('yScale'))
+      .tickValues(this.get('yTicks'))
+      .tickFormat(formatter);
+  }),
+
+  yGridlines: computed('yScale', function() {
+    // The first gridline overlaps the x-axis, so remove it
+    const [, ...ticks] = this.get('yTicks');
```
Comment (meirish) [inline ui/app/components/line-chart.js]: 👌

**Pull request merged**